dbms/src/Formats/CSVRowInputStream.cpp

```diff
@@ -191,46 +162,104 @@ void CSVRowInputStream::readPrefix()
 
-
+/** If you change this function, don't forget to change its counterpart
+  * with extended error reporting: parseRowAndPrintDiagnosticInfo().
+  */
 bool CSVRowInputStream::read(MutableColumns & columns, RowReadExtension & ext)
 {
     if (istr.eof())
         return false;
 
     updateDiagnosticInfo();
 
-    String tmp;
+    /// Track whether we have to fill any columns in this row with default
+    /// values. If not, we return an empty column mask to the caller, so that
+    /// it doesn't have to check it.
+    bool have_default_columns = have_always_default_columns;
+
+    const auto delimiter = format_settings.csv.delimiter;
     for (size_t input_position = 0; input_position < column_indexes_for_input_fields.size(); ++input_position)
     {
         const auto & column_index = column_indexes_for_input_fields[input_position];
         if (column_index)
         {
-            skipWhitespacesAndTabs(istr);
-            data_types[*column_index]->deserializeAsTextCSV(*columns[*column_index], istr, format_settings);
-            skipWhitespacesAndTabs(istr);
+            const auto & type = data_types[*column_index];
+            if (*istr.position() == delimiter
+                    && format_settings.csv.empty_as_default)
+            {
+                /// Treat empty unquoted column value as default value, if
+                /// specified in the settings.
+                read_columns[*column_index] = false;
+                have_default_columns = true;
+            }
+            else
+            {
+                /// Read the column normally.
+                read_columns[*column_index] = true;
+                skipWhitespacesAndTabs(istr);
+                type->deserializeAsTextCSV(*columns[*column_index], istr,
+                    format_settings);
+                skipWhitespacesAndTabs(istr);
+            }
         }
         else
         {
+            /// We never read this column from the file, just skip it.
+            String tmp;
             readCSVString(tmp, istr, format_settings.csv);
         }
 
-        skipDelimiter(istr, format_settings.csv.delimiter, input_position + 1 == column_indexes_for_input_fields.size());
+        skipDelimiter(istr, delimiter,
+            input_position + 1 == column_indexes_for_input_fields.size());
     }
 
-    fillUnreadColumnsWithDefaults(columns, ext);
+    if (have_default_columns)
+    {
+        for (size_t i = 0; i < read_columns.size(); i++)
+        {
+            if (!read_columns[i])
+            {
+                ///!!!FIXME
+                /// 1. Why are the defaults inserted by every row source,
+                /// and not by the caller, given that it walks this array
+                /// anyway?
+                /// 2. The value is overwritten by
+                /// AddingDefaultsBlockInputStream, so we don't care what it
+                /// is. Should we use IColumn::insertDefault() instead?
+                data_types[i]->insertDefaultInto(*columns[i]);
```

Comment:
Does anybody know the answer to the above questions?

Reply:
If I remebmer correctly:

We need to return block with columns of corresponding size. Non-existing values should be prefilled nevertheless. Then we fill default values through applying `if` function or in similar way (that expect block with proper column sizes).

Reply:
(you can update your comment to my)